flang/lib/Lower/OpenMP.cpp

```diff
@@ -716,6 +727,62 @@ genOMP(Fortran::lower::AbstractConverter &converter,
   createBodyOfOp<omp::CriticalOp>(criticalOp, converter, currentLocation, eval);
 }
 
+ void Fortran::lower::genOpenMPSectionsBlock(Fortran::lower::AbstractConverter &converter, Fortran::lower::pft::Evaluation &eval){
+     auto &firOpBuilder = converter.getFirOpBuilder(); 
+     auto currentLocation = converter.getCurrentLocation();
+     mlir::omp::SectionOp sectionOp = firOpBuilder.create<mlir::omp::SectionOp>(currentLocation);
+     createBodyOfOp<omp::SectionOp>(sectionOp, converter, currentLocation, eval);
+ }
+
+ // TODO: Add support for reduction
+ static void
+ genOMP(Fortran::lower::AbstractConverter &converter,
+        Fortran::lower::pft::Evaluation &eval,
+        const Fortran::parser::OpenMPSectionsConstruct &sectionsConstruct) {
+     auto &firOpBuilder = converter.getFirOpBuilder();
+     auto currentLocation = converter.getCurrentLocation();
+     SmallVector<Value> privateClauseOperands, firstPrivateClauseOperands,
+         lastPrivateClauseOperands, reductionVars, allocateOperands,
+         allocatorOperands;
+     mlir::Attribute noWaitClauseOperand;
+     const auto &sectionsClauseList = std::get<Fortran::parser::OmpClauseList>(
+             std::get<Fortran::parser::OmpBeginSectionsDirective>(sectionsConstruct.t).t);
+
+     for(const auto &clause : sectionsClauseList.v){
+
+         if (const auto &lastPrivateClause = std::get_if<Fortran::parser::OmpClause::Lastprivate>(&clause.u)){
+             const Fortran::parser::OmpObjectList &ompObjectList = lastPrivateClause->v;
+             genObjectList(ompObjectList, converter, lastPrivateClauseOperands);
+         }
+
+         if (const auto &allocateClause =
+                 std::get_if<Fortran::parser::OmpClause::Allocate>(&clause.u)) {
+           handleAllocateClause(converter, allocateClause->v, allocatorOperands,
+                                allocateOperands);
+         }
+     }
+     mlir::omp::SectionsOp sectionsOp =
+         firOpBuilder.create<mlir::omp::SectionsOp>(
+             currentLocation, privateClauseOperands, firstPrivateClauseOperands,
+             lastPrivateClauseOperands, reductionVars, nullptr,
+             allocateOperands, allocatorOperands,
+             noWaitClauseOperand.dyn_cast_or_null<UnitAttr>());
+
+     const auto &endSectionsClauseList =
+         std::get<Fortran::parser::OmpEndSectionsDirective>(
+             sectionsConstruct.t);
+     const auto &clauseList =
+         std::get<Fortran::parser::OmpClauseList>(endSectionsClauseList.t);
+     for (const auto &clause : clauseList.v) {
+       if (std::get_if<Fortran::parser::OmpClause::Nowait>(&clause.u)) {
+         sectionsOp.nowaitAttr(firOpBuilder.getUnitAttr());
+       }
+          } 
+     createBodyOfOp<omp::SectionsOp>(sectionsOp, converter, currentLocation, eval);
```

Comment:
createBodyOfOp does privatisation as well (currently only supports private, firstprivate). You can pass the list of clauses as an argument to do this.

Check the following code.
-> https://github.com/flang-compiler/f18-llvm-project/blob/0ee6c0f4a2c05f35e27d3014b9caa8d4ecd12fd0/flang/lib/Lower/OpenMP.cpp#L391
-> https://github.com/flang-compiler/f18-llvm-project/blob/0ee6c0f4a2c05f35e27d3014b9caa8d4ecd12fd0/flang/lib/Lower/OpenMP.cpp#L205